### PR TITLE
Feature/logging updates

### PR DESCRIPTION
**Pull request by benjigoldberg**
Adds a new and improved `log` module. Will phase out `logging` over the next few days.

### CODE REVIEW

**Comment by benjigoldberg** [comment]
/bump minor -- we're still pre-1.0 so I'm not going to rev major since the API isnt yet declared stabled.

File: log/log.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+	} else {
+		logConfig = zap.Config{
+			Level:             zap.NewAtomicLevelAt(level),
+			Development:       false,
+			DisableStacktrace: false,
+			Sampling: &zap.SamplingConfig{
+				Initial:    lc.SamplingInitial,
+				Thereafter: lc.SamplingThereafter,
+			},
+			Encoding:         "json",
```
Comment (briancharous) [inline log/log.go]: Should this be configurable but default to json?

File: log/log.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+			EncoderConfig:    zap.NewProductionEncoderConfig(),
+			OutputPaths:      append(lc.OutputPaths, "stdout"),
+			ErrorOutputPaths: append(lc.ErrorOutputPaths, "stderr"),
+			InitialFields:    map[string]interface{}{"appVersion": lc.AppVersion, "gitSha": lc.GitSha},
```
Comment (briancharous) [inline log/log.go]: I feel like `AppVersion` and `GitSha` should also be optional

File: log/log.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+// useful if you wish for all downstream logs from the site of a given context to include some
+// contextual information. For example, once your application has unpacked the Trace ID, you may
+// wish to log that information with every future request.
+func NewContext(ctx context.Context, fields ...zapcore.Field) context.Context {
+	return context.WithValue(ctx, logKey, Get(ctx).With(fields...))
+}
+
+// Get returns the logger wrapped with the given context. This function is intended to be
+// used as a mechanism for adding scoped arbitrary logging information to the logger. If a nil
+// context is passed, the default global logger is returned.
```
Comment (briancharous) [inline log/log.go]: Hmm... I'd be in favor of killing the global logger altogether

Comment (benjigoldberg) [reply log/log.go]: Why? I think this works nicely and you don't necessarily always have a context to make use of.

Comment (benjigoldberg) [reply log/log.go]: Plus, the hidden logger forces you to use the WithContext interface, so in theory you _should_ always be doing the appropriate thing

File: log/log.go (hunk 2 above)
Comment (benjigoldberg) [reply log/log.go]: Yes, agreed. Optional and supplied as config 👍

File: log/log.go (hunk 1 above)
Comment (benjigoldberg) [reply log/log.go]: agreed, good idea

File: log/log.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+import (
+	"context"
+	"fmt"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+// logKey is the type used to uniquely place the logger within context.Context
```
Comment (greg-mclain) [inline log/log.go]: nitpick, but with the addition of `ctxKey` you may want to either move this comment down a line, or change it from `logKey` to `ctxKey`. We probably don't need a comment on both fields

Comment (benjigoldberg) [reply log/log.go]: nice catch

File: log/log.go [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+// metricsHook is a callback hook used to track logging metrics at runtime
+func (lc *LoggingConfig) metricsHook(entry zapcore.Entry) error {
+	lc.counter.With(prometheus.Labels{"level": entry.Level.CapitalString()}).Inc()
+	return nil
+}
+
+// InitializeLogger sets up the logger. This function should be called as soon
+// as possible. Any use of the logger provided by this package will be a nop
+// until this function is called
```
Comment (wilsoniya) [inline log/log.go]: period

Comment (benjigoldberg) [reply log/log.go]: 👍

File: log/log.go [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+// logKey is the value used to uniquely place the logger within context.Context
+const logKey ctxKey = iota
+
+// logger is the default zap logger
+var logger = zap.NewNop()
+
+// LoggingConfig defines the necessary configuration for instantiating a Logger
+type LoggingConfig struct {
+	UseDevelopmentLogger bool                   // If true, use default zap development logger settings
```
Comment (wilsoniya) [inline log/log.go]: beautiful

File: log/log.go [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+	SamplingThereafter   int                    // Sampling rate after that initial cap is hit, per cycle
+	Encoding             string                 // One of "json" or "console"
+	Fields               map[string]interface{} // These fields will be applied to all logs. Strongly recommend `version` at a minimum.
+	Cores                []zapcore.Core         // Provides optional functionality to allow users to Tee log output to additional Log cores
+	Options              []zap.Option           // Users may specify any additional zap logger options which override defaults
+	counter              *prometheus.CounterVec // Counter tracks the number of logs by level
+}
+
+// metricsHook is a callback hook used to track logging metrics at runtime
+func (lc *LoggingConfig) metricsHook(entry zapcore.Entry) error {
```
Comment (briancharous) [inline log/log.go]: This is pretty nitpicky but I don't think that these receivers need to be pointers

File: log/log_test.go [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			assert.Equal(t, test.expectedOutcome, NewContext(test.ctx, test.fields...))
+		})
+	}
+}
+
+func TestMetricsHook(t *testing.T) {
+	tests := []struct {
```
Comment (briancharous) [inline log/log_test.go]: There's only 1 test case here and the code its testing is like 2 lines long without any conditional branches, personally I'd get rid of the subtest and just write a single unit test in here.

Comment (benjigoldberg) [reply log/log_test.go]: yeah I know, I was being lazy b/c I set this up thinking I'd have more tests. Then, I decided I already set it up so I could just leave it for the next person.

File: log/log.go (hunk 3 above)
Comment (briancharous) [reply log/log.go]: Yeah I guess this is fine. The alternative would be either erroring out or silently not logging, both of which are worse than a global.

**Pull request merged**